### PR TITLE
Added a perfetto connector

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
So Google is developing this tool called Perfetto (h/t @jrmadsen). It's a really nice trace viewer, and they distribute it as single-file SDK, like SQLite. So it's a simple tool to distribute, but really fully featured. This just connects it to Kokkos

### CODE REVIEW

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+extern "C" void kokkosp_begin_parallel_scan(const char* name,
+                                            const uint32_t,
+                                            uint64_t*);
+extern "C" void kokkosp_end_parallel_scan(const uint64_t);
+extern "C" void kokkosp_begin_parallel_reduce(const char* name,
+                                              const uint32_t,
+                                              uint64_t*);
+extern "C" void kokkosp_end_parallel_reduce(const uint64_t);
+extern "C" void kokkosp_push_profile_region(
+    char* regionName);
```
Comment (nliber) [inline profiling/perfetto-connector/libperfetto-connector.cc:22]: Is there a reason that `regionName` isn't a `const char *`?

**Review by nliber — CHANGES_REQUESTED** [review]

File: profiling/perfetto-connector/libperfetto-connector.cc (hunk 1 above)
Comment (nliber) [reply profiling/perfetto-connector/libperfetto-connector.cc:22]: Also, I only looked at libperfetto-connector.cc

Comment (DavidPoliakoff) [reply profiling/perfetto-connector/libperfetto-connector.cc:22]: Purely out of consistency with other tools. It could be const char*

**Comment by DavidPoliakoff** [comment]
Also pinging @daboehme, as this is pretty close to what he'd want to do to support Perfetto in Caliper

File: profiling/perfetto-connector/libperfetto-connector.cc (hunk 1 above)
Comment (nliber) [reply profiling/perfetto-connector/libperfetto-connector.cc:22]: There may be some historical C reasons behind it.  Anyway, I'll approve.

**Review by nliber — APPROVED** [review]

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+extern "C" void kokkosp_end_parallel_scan(const uint64_t) {
+  TRACE_EVENT_END("kokkos.kernels");
+}
+
+extern "C" void kokkosp_begin_parallel_reduce(const char* name,
+                                              const uint32_t,
+                                              uint64_t*) {
+  TRACE_EVENT_BEGIN("kokkos.kernels", perfetto::StaticString(name));
```
Comment (jrmadsen) [inline profiling/perfetto-connector/libperfetto-connector.cc:97]: Don't ever look behind their macros. I tried making the "kokkos.kernels" part something a downstream user could set and they are unbelievably massive and convoluted.

Comment (daboehme) [reply profiling/perfetto-connector/libperfetto-connector.cc:97]: Good to know 😄

Comment (DavidPoliakoff) [reply profiling/perfetto-connector/libperfetto-connector.cc:97]: Via the Perfetto folks, apparently you can do that this way: https://perfetto.dev/docs/instrumentation/track-events#dynamic-and-test-only-categories

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+PERFETTO_DEFINE_CATEGORIES(
+    perfetto::Category("kokkos.kernels")
+        .SetDescription("Kokkos Kernel Events"),
+    perfetto::Category("kokkos.regions").SetDescription(
+        "Kokkos Region Events"));
+PERFETTO_TRACK_EVENT_STATIC_STORAGE();
+perfetto::TraceConfig cfg;
+std::unique_ptr<perfetto::TracingSession> tracing_session;
+int fd;
```
Comment (dalg24) [inline profiling/perfetto-connector/libperfetto-connector.cc:34]: I would prefer if you prefixed these global variables with something like `kokkosp_perfetto_`

Comment (dalg24) [reply profiling/perfetto-connector/libperfetto-connector.cc:34]: Actually `cfg` is being shadowed in `kokkosp_init_library` which is the only place that was using a trace config in the first place.  Remove the global variable.

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  cfg.add_buffers()->set_size_kb(1024);  // Record up to 1 MiB.
+  auto* ds_cfg = cfg.add_data_sources()->mutable_config();
+  perfetto::protos::gen::TrackEventConfig track_event_cfg;
+  //track_event_cfg.add_disabled_categories("*");
+  track_event_cfg.add_enabled_categories("kokkos.kernels");
+  track_event_cfg.add_enabled_categories("kokkos.regions");
+  ds_cfg->set_name("track_event");
+  ds_cfg->set_track_event_config_raw(track_event_cfg.SerializeAsString());
+  tracing_session = 
+        	      std::move(perfetto::Tracing::NewTrace());
```
Comment (dalg24) [inline profiling/perfetto-connector/libperfetto-connector.cc:61]: Fix format.  Also pretty sure you don't need the std::move here.

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                                              const uint32_t,
+                                              uint64_t*) {
+  TRACE_EVENT_BEGIN("kokkos.kernels", perfetto::StaticString(name));
+}
+
+extern "C" void kokkosp_end_parallel_reduce(const uint64_t) {
+  TRACE_EVENT_END("kokkos.kernels");
+}
+
+extern "C" void kokkosp_push_profile_region(char* regionName) {
```
Comment (dalg24) [inline profiling/perfetto-connector/libperfetto-connector.cc:104]: missing `const`?

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: profiling/perfetto-connector/libperfetto-connector.cc (hunk 1 above)
Comment (dalg24) [reply profiling/perfetto-connector/libperfetto-connector.cc:22]: https://github.com/kokkos/kokkos-tools/wiki/Profiling-Hooks#kokkosp_push_profile_region

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+
+PERFETTO_DEFINE_CATEGORIES(
+    perfetto::Category("kokkos.kernels")
+        .SetDescription("Kokkos Kernel Events"),
+    perfetto::Category("kokkos.regions").SetDescription(
+        "Kokkos Region Events"));
+PERFETTO_TRACK_EVENT_STATIC_STORAGE();
+perfetto::TraceConfig cfg;
+std::unique_ptr<perfetto::TracingSession> tracing_session;
```
Comment (jrmadsen) [inline profiling/perfetto-connector/libperfetto-connector.cc]:
```suggestion
std::unique_ptr<perfetto::TracingSession> kokkosp_perfetto_tracing_session;
```

File: profiling/perfetto-connector/libperfetto-connector.cc (hunk 4 above)
Comment (jrmadsen) [inline profiling/perfetto-connector/libperfetto-connector.cc]:
```suggestion
  kokkosp_perfetto_tracing_session = perfetto::Tracing::NewTrace();
```

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  perfetto::protos::gen::TrackEventConfig track_event_cfg;
+  //track_event_cfg.add_disabled_categories("*");
+  track_event_cfg.add_enabled_categories("kokkos.kernels");
+  track_event_cfg.add_enabled_categories("kokkos.regions");
+  ds_cfg->set_name("track_event");
+  ds_cfg->set_track_event_config_raw(track_event_cfg.SerializeAsString());
+  tracing_session = 
+        	      std::move(perfetto::Tracing::NewTrace());
+  tracing_session->Setup(cfg, fd);
+  tracing_session->StartBlocking();
```
Comment (jrmadsen) [inline profiling/perfetto-connector/libperfetto-connector.cc]:
```suggestion
  kokkosp_perfetto_tracing_session->Setup(cfg, fd);
  kokkosp_perfetto_tracing_session->StartBlocking();
```

File: profiling/perfetto-connector/libperfetto-connector.cc (hunk 5 above)
Comment (jrmadsen) [inline profiling/perfetto-connector/libperfetto-connector.cc]:
```suggestion
extern "C" void kokkosp_push_profile_region(const char* regionName) {
```

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+extern "C" void kokkosp_finalize_library() {
+  printf("-----------------------------------------------------------\n");
+  printf("KokkosP: Finalization of Perfetto Connector. Complete.\n");
+  printf("-----------------------------------------------------------\n");
+  tracing_session->StopBlocking();
```
Comment (jrmadsen) [inline profiling/perfetto-connector/libperfetto-connector.cc]:
```suggestion
  kokkosp_perfetto_tracing_session->StopBlocking();
```

**Review by jrmadsen — COMMENTED** [review]
Just adding suggestions to streamline this

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+extern "C" void kokkosp_pop_profile_region();
+
+
+PERFETTO_DEFINE_CATEGORIES(
+    perfetto::Category("kokkos.kernels")
+        .SetDescription("Kokkos Kernel Events"),
+    perfetto::Category("kokkos.regions").SetDescription(
+        "Kokkos Region Events"));
+PERFETTO_TRACK_EVENT_STATIC_STORAGE();
+perfetto::TraceConfig cfg;
```
Comment (jrmadsen) [inline profiling/perfetto-connector/libperfetto-connector.cc]:
```suggestion
```

File: profiling/perfetto-connector/libperfetto-connector.cc [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+extern "C" void kokkosp_begin_parallel_scan(const char* name,
+                                            const uint32_t,
+                                            uint64_t*);
+extern "C" void kokkosp_end_parallel_scan(const uint64_t);
+extern "C" void kokkosp_begin_parallel_reduce(const char* name,
+                                              const uint32_t,
+                                              uint64_t*);
+extern "C" void kokkosp_end_parallel_reduce(const uint64_t);
+extern "C" void kokkosp_push_profile_region(
+    char* regionName);
```
Comment (jrmadsen) [inline profiling/perfetto-connector/libperfetto-connector.cc]:
```suggestion
    const char* regionName);
```

**Review by jrmadsen — APPROVED** [review]

**Comment by crtrott** [comment]
closed in favor of #142

**Pull request closed without merging**